File: tests/io_operations/csv_processing/csv_processing.py
#!/usr/bin/env python3
"""
CSV processing test implementation in Python.
Measures CSV file read, write, filtering, and aggregation performance.
Optimized version for better performance.
"""

import json
import sys
import time
import random
import string
from typing import Dict, List, Any, Union


def load_test_data() -> List[List[str]]:
    """Load standardized test data from JSON file."""
    import os
    test_data_path = os.path.join(os.path.dirname(__file__), 'test_data.json')
    with open(test_data_path, 'r') as f:
        test_data = json.load(f)
    
    csv_data = test_data['csv_data']
    data = [csv_data['headers']]
    data.extend(csv_data['rows'])
    return data

def generate_csv_data(rows: int, cols: int, data_type: str) -> List[List[str]]:
    """Generate CSV data using standardized test data - replicated to match size."""
    base_data = load_test_data()
    
    # Use the headers from test data
    headers = base_data[0][:cols] if cols <= len(base_data[0]) else base_data[0] + [f"col_{i+1}" for i in range(len(base_data[0]), cols)]
    data = [headers]
    
    # Replicate base rows to match requested size
    base_rows = base_data[1:]
    for row_idx in range(rows):
        source_row = base_rows[row_idx % len(base_rows)]
        row_data = source_row[:cols] if cols <= len(source_row) else source_row + [f"extra_{row_idx}_{i}" for i in range(len(source_row), cols)]
        data.append(row_data)
    
    return data


def write_csv_to_string(data: List[List[str]]) -> str:
    """Write CSV data to string format using optimized string operations."""
    # Use more efficient string building
    result = []
    for row in data:
        result.append(','.join(row))
    return '\n'.join(result) + '\n'


def read_csv_from_string(csv_string: str) -> List[List[str]]:
    """Read CSV data from string format using optimized string operations."""
    # More efficient parsing
    lines = csv_string.strip().split('\n')
    return [line.split(',') for line in lines if line]


def filter_csv_data(data: List[List[str]], filter_column: int = 0) -> List[List[str]]:
    """Filter CSV data based on a condition - optimized."""
    if not data or len(data) < 2:
        return data
    
    filtered_data = [data[0]]  # Headers
    
    for row in data[1:]:
        if len(row) > filter_column:
            cell_value = row[filter_column]
            # Quick numeric check without exception handling
            if cell_value.replace('.', '').replace('-', '').isdigit():
                if float(cell_value) > 500:
                    filtered_data.append(row)
            elif len(cell_value) > 5:
                filtered_data.append(row)
    return filtered_data


def aggregate_csv_data(data: List[List[str]]) -> Dict[str, Any]:
    """Perform aggregation operations on CSV data - optimized."""
    if not data or len(data) < 2:
        return {}
    
    headers = data[0]
    numeric_columns = []
    
    # Quick numeric column detection
    for col_idx in range(len(headers)):
        is_numeric = True
        check_rows = min(3, len(data) - 1)  # Check fewer rows
        for row_idx in range(1, check_rows + 1):
            if len(data[row_idx]) > col_idx:
                cell = data[row_idx][col_idx]
                if not (cell.replace('.', '').replace('-', '').isdigit()):
                    is_numeric = False
                    break
        if is_numeric:
            numeric_columns.append(col_idx)
    
    aggregations = {}
    
    for col_idx in numeric_columns:
        col_name = headers[col_idx]
        values = []
        
        for row in data[1:]:
            if len(row) > col_idx:
                cell = row[col_idx]
                if cell.replace('.', '').replace('-', '').isdigit():
                    values.append(float(cell))
        
        if values:
            total = sum(values)
            aggregations[col_name] = {
                "sum": total,
                "avg": total / len(values),
                "min": min(values),
                "max": max(values),
                "count": len(values)
            }
    
    return aggregations


def run_csv_processing_benchmark(config: Dict) -> Dict:
    """Run CSV processing benchmark."""
    row_counts = config.get("row_counts", [1000])
    column_counts = config.get("column_counts", [5])
    operations = config.get("operations", ["read", "write"])
    data_types = config.get("data_types", ["mixed"])
    iterations = config.get("iterations", 3)
    
    results = {
        "start_time": time.time(),
        "test_cases": [],
        "summary": {
            "total_tests": 0,
            "successful_tests": 0,
            "failed_tests": 0,
            "avg_read_time": 0.0,
            "avg_write_time": 0.0,
            "avg_filter_time": 0.0,
            "avg_aggregate_time": 0.0
        }
    }
    
    all_read_times = []
    all_write_times = []
    all_filter_times = []
    all_aggregate_times = []
    
    for rows in row_counts:
        for cols in column_counts:
            for data_type in data_types:
                print(f"Testing CSV: {rows} rows x {cols} cols, type: {data_type}...", file=sys.stderr)
                
                test_case = {
                    "row_count": rows,
                    "column_count": cols,
                    "data_type": data_type,
                    "operations": operations,
                    "iterations": [],
                    "avg_read_time": 0.0,
                    "avg_write_time": 0.0,
                    "avg_filter_time": 0.0,
                    "avg_aggregate_time": 0.0
                }
                
                read_times = []
                write_times = []
                filter_times = []
                aggregate_times = []
                
                for i in range(iterations):
                    print(f"  Iteration {i+1}/{iterations}...", file=sys.stderr)
                    
                    # Generate test data
                    csv_data = generate_csv_data(rows, cols, data_type)
                    
                    iteration_result = {
                        "iteration": i + 1,
                        "data_size": len(csv_data),
                        "operations": {}
                    }
                    
                    results["summary"]["total_tests"] += 1
                    success = True
                    
                    # Write operation
                    if "write" in operations:
                        try:
                            start_time = time.time()
                            csv_string = write_csv_to_string(csv_data)
                            write_time = (time.time() - start_time) * 1000  # ms
                            
                            write_times.append(write_time)
                            all_write_times.append(write_time)
                            
                            iteration_result["operations"]["write"] = {
                                "success": True,
                                "time_ms": write_time,
                                "output_size": len(csv_string)
                            }
                            
                        except Exception as e:
                            success = False
                            iteration_result["operations"]["write"] = {
                                "success": False,
                                "error": str(e)
                            }
                    
                    # Read operation
                    if "read" in operations:
                        try:
                            # First write to get string representation
                            csv_string = write_csv_to_string(csv_data)
                            
                            start_time = time.time()
                            read_data = read_csv_from_string(csv_string)
                            read_time = (time.time() - start_time) * 1000  # ms
                            
                            read_times.append(read_time)
                            all_read_times.append(read_time)
                            
                            iteration_result["operations"]["read"] = {
                                "success": True,
                                "time_ms": read_time,
                                "rows_read": len(read_data)
                            }
                            
                        except Exception as e:
                            success = False
                            iteration_result["operations"]["read"] = {
                                "success": False,
                                "error": str(e)
                            }
                    
                    # Filter operation
                    if "filter" in operations:
                        try:
                            start_time = time.time()
                            filtered_data = filter_csv_data(csv_data)
                            filter_time = (time.time() - start_time) * 1000  # ms
                            
                            filter_times.append(filter_time)
                            all_filter_times.append(filter_time)
                            
                            iteration_result["operations"]["filter"] = {
                                "success": True,
                                "time_ms": filter_time,
                                "original_rows": len(csv_data),
                                "filtered_rows": len(filtered_data)
                            }
                            
                        except Exception as e:
                            success = False
                            iteration_result["operations"]["filter"] = {
                                "success": False,
                                "error": str(e)
                            }
                    
                    # Aggregate operation
                    if "aggregate" in operations:
                        try:
                            start_time = time.time()
                            aggregations = aggregate_csv_data(csv_data)
                            aggregate_time = (time.time() - start_time) * 1000  # ms
                            
                            aggregate_times.append(aggregate_time)
                            all_aggregate_times.append(aggregate_time)
                            
                            iteration_result["operations"]["aggregate"] = {
                                "success": True,
                                "time_ms": aggregate_time,
                                "aggregated_columns": len(aggregations)
                            }
                            
                        except Exception as e:
                            success = False
                            iteration_result["operations"]["aggregate"] = {
                                "success": False,
                                "error": str(e)
                            }
                    
                    if success:
                        results["summary"]["successful_tests"] += 1
                    else:
                        results["summary"]["failed_tests"] += 1
                    
                    test_case["iterations"].append(iteration_result)
                
                # Calculate averages for this test case
                if read_times:
                    test_case["avg_read_time"] = sum(read_times) / len(read_times)
                if write_times:
                    test_case["avg_write_time"] = sum(write_times) / len(write_times)
                if filter_times:
                    test_case["avg_filter_time"] = sum(filter_times) / len(filter_times)
                if aggregate_times:
                    test_case["avg_aggregate_time"] = sum(aggregate_times) / len(aggregate_times)
                
                results["test_cases"].append(test_case)
    
    # Calculate overall summary
    if all_read_times:
        results["summary"]["avg_read_time"] = sum(all_read_times) / len(all_read_times)
    if all_write_times:
        results["summary"]["avg_write_time"] = sum(all_write_times) / len(all_write_times)
    if all_filter_times:
        results["summary"]["avg_filter_time"] = sum(all_filter_times) / len(all_filter_times)
    if all_aggregate_times:
        results["summary"]["avg_aggregate_time"] = sum(all_aggregate_times) / len(all_aggregate_times)
    
    results["end_time"] = time.time()
    results["total_execution_time"] = results["end_time"] - results["start_time"]
    
    return results


def main():
    """Main entry point for CSV processing test."""
    if len(sys.argv) < 2:
        print("Usage: python csv_processing.py <config_file>", file=sys.stderr)
        sys.exit(1)
    
    config_file = sys.argv[1]
    
    try:
        with open(config_file, 'r') as f:
            config = json.load(f)
        
        parameters = config.get("parameters", {})
        results = run_csv_processing_benchmark(parameters)
        
        print(json.dumps(results, indent=2))
        
    except FileNotFoundError:
        print(f"Error: Config file '{config_file}' not found", file=sys.stderr)
        sys.exit(1)
    except json.JSONDecodeError as e:
        print(f"Error: Invalid JSON in config file: {e}", file=sys.stderr)
        sys.exit(1)
    except Exception as e:
        print(f"Error: {e}", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()

File: tests/io_operations/csv_processing/csv_processing.ts
#!/usr/bin/env node

/**
 * CSV processing test implementation in TypeScript.
 * Measures CSV file read, write, filtering, and aggregation performance.
 */

import * as fs from 'fs';
import * as process from 'process';

interface Config {
    parameters: {
        row_counts?: number[];
        column_counts?: number[];
        operations?: string[];
        data_types?: string[];
        iterations?: number;
    };
}

interface OperationResult {
    success: boolean;
    time_ms?: number;
    error?: string;
    output_size?: number;
    rows_read?: number;
    original_rows?: number;
    filtered_rows?: number;
    aggregated_columns?: number;
}

interface IterationResult {
    iteration: number;
    data_size: number;
    operations: { [key: string]: OperationResult };
}

interface TestCase {
    row_count: number;
    column_count: number;
    data_type: string;
    operations: string[];
    iterations: IterationResult[];
    avg_read_time: number;
    avg_write_time: number;
    avg_filter_time: number;
    avg_aggregate_time: number;
}

interface Summary {
    total_tests: number;
    successful_tests: number;
    failed_tests: number;
    avg_read_time: number;
    avg_write_time: number;
    avg_filter_time: number;
    avg_aggregate_time: number;
}

interface Results {
    start_time: number;
    test_cases: TestCase[];
    summary: Summary;
    end_time: number;
    total_execution_time: number;
}

function generateCSVData(rows: number, cols: number, dataType: string): string[][] {
    const data: string[][] = [];
    
    // Generate headers efficiently
    const headers: string[] = Array.from({ length: cols }, (_, i) => `col_${i + 1}`);
    data.push(headers);
    
    // Pre-generate values for better performance
    const numericValues = Array.from({ length: 1000 }, (_, i) => (i * 1.5 + 100).toFixed(2));
    const textValues = Array.from({ length: 100 }, (_, i) => `text_${i}_data`);
    
    // Generate data rows efficiently
    for (let row = 0; row < rows; row++) {
        const rowData: string[] = [];
        for (let col = 0; col < cols; col++) {
            let value: string;
            
            if (dataType === "numeric") {
                value = numericValues[row % numericValues.length];
            } else if (dataType === "text") {
                value = textValues[col % textValues.length] + `_${row}`;
            } else { // mixed
                switch (col % 3) {
                    case 0:
                        value = String(row * 10 + col);
                        break;
                    case 1:
                        value = `item_${row}_${col}`;
                        break;
                    default:
                        value = ((row + col) * 1.5).toFixed(2);
                        break;
                }
            }
            
            rowData.push(value);
        }
        data.push(rowData);
    }
    
    return data;
}

function writeCSVToString(data: string[][]): string {
    return data.map(row => row.join(',')).join('\n') + '\n';
}

function readCSVFromString(csvString: string): string[][] {
    return csvString.trim().split('\n').map(line => line.split(','));
}

function filterCSVData(data: string[][], filterColumn: number = 0): string[][] {
    if (data.length < 2) {
        return data;
    }
    
    const filtered: string[][] = [data[0]]; // Keep headers
    
    for (let i = 1; i < data.length; i++) {
        const row = data[i];
        if (row.length > filterColumn) {
            const cellValue = row[filterColumn];
            // Quick numeric check without parseFloat
            const isNumeric = /^-?\d*\.?\d+$/.test(cellValue);
            if (isNumeric) {
                if (parseFloat(cellValue) > 500) {
                    filtered.push(row);
                }
            } else if (cellValue.length > 5) {
                filtered.push(row);
            }
        }
    }
    
    return filtered;
}

function aggregateCSVData(data: string[][]): { [key: string]: { [key: string]: number } } {
    if (data.length < 2) {
        return {};
    }
    
    const headers = data[0];
    const numericColumns: number[] = [];
    
    // Find numeric columns (check first 3 rows for efficiency)
    for (let colIdx = 0; colIdx < headers.length; colIdx++) {
        let isNumeric = true;
        const checkRows = Math.min(3, data.length - 1);
        
        for (let rowIdx = 1; rowIdx <= checkRows; rowIdx++) {
            if (colIdx < data[rowIdx].length) {
                const cellValue = data[rowIdx][colIdx];
                if (!/^-?\d*\.?\d+$/.test(cellValue)) {
                    isNumeric = false;
                    break;
                }
            }
        }
        if (isNumeric) {
            numericColumns.push(colIdx);
        }
    }
    
    const aggregations: { [key: string]: { [key: string]: number } } = {};
    
    for (const colIdx of numericColumns) {
        const colName = headers[colIdx];
        let sum = 0;
        let min = Infinity;
        let max = -Infinity;
        let count = 0;
        
        // Single pass aggregation for efficiency
        for (let i = 1; i < data.length; i++) {
            const row = data[i];
            if (colIdx < row.length) {
                const value = parseFloat(row[colIdx]);
                if (!isNaN(value)) {
                    sum += value;
                    min = Math.min(min, value);
                    max = Math.max(max, value);
                    count++;
                }
            }
        }
        
        if (count > 0) {
            aggregations[colName] = {
                sum: sum,
                avg: sum / count,
                min: min,
                max: max,
                count: count
            };
        }
    }
    
    return aggregations;
}

function runCSVProcessingBenchmark(config: Config): Results {
    const parameters = config.parameters;
    
    // Set defaults
    const rowCounts = parameters.row_counts || [1000];
    const columnCounts = parameters.column_counts || [5];
    const operations = parameters.operations || ["read", "write"];
    const dataTypes = parameters.data_types || ["mixed"];
    const iterations = parameters.iterations || 3;
    
    const startTime = Date.now();
    const testCases: TestCase[] = [];
    const allReadTimes: number[] = [];
    const allWriteTimes: number[] = [];
    const allFilterTimes: number[] = [];
    const allAggregateTimes: number[] = [];
    let totalTests = 0;
    let successfulTests = 0;
    let failedTests = 0;
    
    for (const rows of rowCounts) {
        for (const cols of columnCounts) {
            for (const dataType of dataTypes) {
                console.error(`Testing CSV: ${rows} rows x ${cols} cols, type: ${dataType}...`);
                
                const readTimes: number[] = [];
                const writeTimes: number[] = [];
                const filterTimes: number[] = [];
                const aggregateTimes: number[] = [];
                const iterationsData: IterationResult[] = [];
                
                for (let i = 0; i < iterations; i++) {
                    console.error(`  Iteration ${i + 1}/${iterations}...`);
                    
                    // Generate test data
                    const csvData = generateCSVData(rows, cols, dataType);
                    
                    const iterationResult: IterationResult = {
                        iteration: i + 1,
                        data_size: csvData.length,
                        operations: {}
                    };
                    
                    totalTests++;
                    let success = true;
                    
                    // Write operation
                    if (operations.includes("write")) {
                        try {
                            const start = process.hrtime.bigint();
                            const csvString = writeCSVToString(csvData);
                            const writeTime = Number(process.hrtime.bigint() - start) / 1000000; // Convert to ms
                            
                            writeTimes.push(writeTime);
                            allWriteTimes.push(writeTime);
                            
                            iterationResult.operations["write"] = {
                                success: true,
                                time_ms: writeTime,
                                output_size: csvString.length
                            };
                        } catch (e) {
                            success = false;
                            iterationResult.operations["write"] = {
                                success: false,
                                error: String(e)
                            };
                        }
                    }
                    
                    // Read operation
                    if (operations.includes("read")) {
                        try {
                            const csvString = writeCSVToString(csvData);
                            
                            const start = process.hrtime.bigint();
                            const readData = readCSVFromString(csvString);
                            const readTime = Number(process.hrtime.bigint() - start) / 1000000; // Convert to ms
                            
                            readTimes.push(readTime);
                            allReadTimes.push(readTime);
                            
                            iterationResult.operations["read"] = {
                                success: true,
                                time_ms: readTime,
                                rows_read: readData.length
                            };
                        } catch (e) {
                            success = false;
                            iterationResult.operations["read"] = {
                                success: false,
                                error: String(e)
                            };
                        }
                    }
                    
                    // Filter operation
                    if (operations.includes("filter")) {
                        try {
                            const start = process.hrtime.bigint();
                            const filteredData = filterCSVData(csvData);
                            const filterTime = Number(process.hrtime.bigint() - start) / 1000000; // Convert to ms
                            
                            filterTimes.push(filterTime);
                            allFilterTimes.push(filterTime);
                            
                            iterationResult.operations["filter"] = {
                                success: true,
                                time_ms: filterTime,
                                original_rows: csvData.length,
                                filtered_rows: filteredData.length
                            };
                        } catch (e) {
                            success = false;
                            iterationResult.operations["filter"] = {
                                success: false,
                                error: String(e)
                            };
                        }
                    }
                    
                    // Aggregate operation
                    if (operations.includes("aggregate")) {
                        try {
                            const start = process.hrtime.bigint();
                            const aggregations = aggregateCSVData(csvData);
                            const aggregateTime = Number(process.hrtime.bigint() - start) / 1000000; // Convert to ms
                            
                            aggregateTimes.push(aggregateTime);
                            allAggregateTimes.push(aggregateTime);
                            
                            iterationResult.operations["aggregate"] = {
                                success: true,
                                time_ms: aggregateTime,
                                aggregated_columns: Object.keys(aggregations).length
                            };
                        } catch (e) {
                            success = false;
                            iterationResult.operations["aggregate"] = {
                                success: false,
                                error: String(e)
                            };
                        }
                    }
                    
                    if (success) {
                        successfulTests++;
                    } else {
                        failedTests++;
                    }
                    
                    iterationsData.push(iterationResult);
                }
                
                // Calculate averages for this test case
                const testCase: TestCase = {
                    row_count: rows,
                    column_count: cols,
                    data_type: dataType,
                    operations: operations,
                    iterations: iterationsData,
                    avg_read_time: readTimes.length > 0 ? readTimes.reduce((a, b) => a + b, 0) / readTimes.length : 0,
                    avg_write_time: writeTimes.length > 0 ? writeTimes.reduce((a, b) => a + b, 0) / writeTimes.length : 0,
                    avg_filter_time: filterTimes.length > 0 ? filterTimes.reduce((a, b) => a + b, 0) / filterTimes.length : 0,
                    avg_aggregate_time: aggregateTimes.length > 0 ? aggregateTimes.reduce((a, b) => a + b, 0) / aggregateTimes.length : 0
                };
                
                testCases.push(testCase);
            }
        }
    }
    
    // Calculate overall summary
    const summary: Summary = {
        total_tests: totalTests,
        successful_tests: successfulTests,
        failed_tests: failedTests,
        avg_read_time: allReadTimes.length > 0 ? allReadTimes.reduce((a, b) => a + b, 0) / allReadTimes.length : 0,
        avg_write_time: allWriteTimes.length > 0 ? allWriteTimes.reduce((a, b) => a + b, 0) / allWriteTimes.length : 0,
        avg_filter_time: allFilterTimes.length > 0 ? allFilterTimes.reduce((a, b) => a + b, 0) / allFilterTimes.length : 0,
        avg_aggregate_time: allAggregateTimes.length > 0 ? allAggregateTimes.reduce((a, b) => a + b, 0) / allAggregateTimes.length : 0
    };
    
    const endTime = Date.now();
    const totalExecutionTime = (endTime - startTime) / 1000;
    
    return {
        start_time: startTime / 1000,
        test_cases: testCases,
        summary: summary,
        end_time: endTime / 1000,
        total_execution_time: totalExecutionTime
    };
}

function main(): void {
    if (process.argv.length < 3) {
        console.error("Usage: node csv_processing.js <config_file>");
        process.exit(1);
    }
    
    const configFile = process.argv[2];
    
    try {
        const configContent = fs.readFileSync(configFile, 'utf8');
        const config: Config = JSON.parse(configContent);
        
        const results = runCSVProcessingBenchmark(config);
        
        console.log(JSON.stringify(results, null, 2));
    } catch (e) {
        if (e instanceof Error) {
            if ((e as any).code === 'ENOENT') {
                console.error(`Error: Config file '${configFile}' not found`);
            } else if (e instanceof SyntaxError) {
                console.error(`Error: Invalid JSON in config file: ${e.message}`);
            } else {
                console.error(`Error: ${e.message}`);
            }
        } else {
            console.error(`Error: ${String(e)}`);
        }
        process.exit(1);
    }
}

// Always run main for benchmark purposes
main();